File: src/testing/command-harness.ts
import {
  CommandRegistry,
  createCommandRegistry,
  type CommandContext,
  type CommandSpec,
  type CreateCommandRegistryOptions,
} from '../commands/index.js';
import { resolveExecutionPolicy } from '../execution-policy.js';
import { SimpleMemory } from '../memory.js';
import { textDecoder, textEncoder, type CommandResult } from '../types.js';
import { MemoryVFS } from '../vfs/memory-vfs.js';

export type CreateTestCommandContextOptions = Partial<CommandContext>;

export interface RunRegisteredCommandOptions {
  ctx?: CommandContext;
  stdin?: Uint8Array;
}

export interface RunRegisteredCommandResult {
  ctx: CommandContext;
  result: CommandResult;
  spec: CommandSpec;
}

export const NO_STDIN = new Uint8Array();

export function createTestCommandRegistry(options: CreateCommandRegistryOptions = {}): CommandRegistry {
  return createCommandRegistry(options);
}

export function createTestCommandContext(overrides: CreateTestCommandContextOptions = {}): CommandContext {
  const registry = overrides.registry ?? createTestCommandRegistry();

  return {
    vfs: overrides.vfs ?? new MemoryVFS(),
    adapters: overrides.adapters ?? {},
    memory: overrides.memory ?? new SimpleMemory(),
    registry,
    executionPolicy: overrides.executionPolicy ?? resolveExecutionPolicy(),
    outputDir: overrides.outputDir ?? '/output',
    outputCounter: overrides.outputCounter ?? 0,
  };
}

export function getRegisteredCommand(name: string, registry: CommandRegistry): CommandSpec {
  const spec = registry.get(name);
  if (!spec) {
    throw new Error(`expected command to exist: ${name}`);
  }
  return spec;
}

export async function runRegisteredCommand(
  name: string,
  args: string[] = [],
  options: RunRegisteredCommandOptions = {},
): Promise<RunRegisteredCommandResult> {
  const ctx = options.ctx ?? createTestCommandContext();
  const spec = getRegisteredCommand(name, ctx.registry);
  const result = await spec.handler(ctx, [...args], options.stdin ?? NO_STDIN);

  return { ctx, result, spec };
}

export function stdoutText(result: CommandResult | Uint8Array): string {
  return textDecoder.decode(result instanceof Uint8Array ? result : result.stdout);
}

export function stdinText(text: string): Uint8Array {
  return textEncoder.encode(text);
}

File: python/src/onetool/testing/command_harness.py
from __future__ import annotations

from dataclasses import dataclass

from ..commands import CommandContext, CommandRegistry, CommandSpec, create_command_registry, maybe_await
from ..commands.register import CreateCommandRegistryOptions
from ..execution_policy import AgentCLIExecutionPolicy, resolve_execution_policy
from ..memory import SimpleMemory
from ..types import CommandResult, ToolAdapters
from ..vfs import VFS, MemoryVFS

NO_STDIN = b""


@dataclass(frozen=True, slots=True)
class CreateTestCommandContextOptions:
    vfs: VFS | None = None
    adapters: ToolAdapters | None = None
    memory: SimpleMemory | None = None
    registry: CommandRegistry | None = None
    execution_policy: AgentCLIExecutionPolicy | None = None
    output_dir: str | None = None
    output_counter: int | None = None


@dataclass(frozen=True, slots=True)
class RunRegisteredCommandOptions:
    ctx: CommandContext | None = None
    stdin: bytes = NO_STDIN


@dataclass(frozen=True, slots=True)
class RunRegisteredCommandResult:
    ctx: CommandContext
    result: CommandResult
    spec: CommandSpec


def create_test_command_registry(options: CreateCommandRegistryOptions | None = None) -> CommandRegistry:
    return create_command_registry(options)


def create_test_command_context(
    options: CreateTestCommandContextOptions | None = None,
) -> CommandContext:
    resolved = options or CreateTestCommandContextOptions()
    registry = resolved.registry or create_test_command_registry()
    execution_policy = resolve_execution_policy(resolved.execution_policy)

    return CommandContext(
        vfs=resolved.vfs or MemoryVFS(),
        adapters=resolved.adapters or ToolAdapters(),
        memory=resolved.memory or SimpleMemory(),
        registry=registry,
        execution_policy=execution_policy,
        output_dir=resolved.output_dir or "/output",
        output_counter=resolved.output_counter or 0,
    )


def get_registered_command(name: str, registry: CommandRegistry) -> CommandSpec:
    spec = registry.get(name)
    if spec is None:
        raise ValueError(f"expected command to exist: {name}")
    return spec


async def run_registered_command(
    name: str,
    args: list[str] | tuple[str, ...] | None = None,
    options: RunRegisteredCommandOptions | None = None,
) -> RunRegisteredCommandResult:
    resolved = options or RunRegisteredCommandOptions()
    ctx = resolved.ctx or create_test_command_context()
    spec = get_registered_command(name, ctx.registry)
    result = await maybe_await(spec.handler(ctx, list(args or ()), resolved.stdin))

    return RunRegisteredCommandResult(
        ctx=ctx,
        result=result,
        spec=spec,
    )


def stdout_text(result: CommandResult | bytes) -> str:
    data = result if isinstance(result, bytes) else result.stdout
    return data.decode("utf-8")


def stdin_text(text: str) -> bytes:
    return text.encode("utf-8")
